lib/inets/src/http_server/mod_esi.erl

```diff
@@ -220,47 +220,47 @@ erl(#mod{method = "POST", entity_body = Body} = ModData, ESIBody, Modules) ->
-generate_webpage(ModData, ESIBody, Modules, Module, Function,
-		 Input, ScriptElements) when is_atom(Module), is_atom(Function) ->
-    case lists:member(Module, Modules) of
-	true ->
-	    Env = httpd_script_env:create_env(esi, ModData, ScriptElements),
-	    case erl_scheme_webpage_chunk(Module, Function, 
-					  Env, Input, ModData) of
-		{error, erl_scheme_webpage_chunk_undefined} ->
-                    {proceed, [{status, {404, ModData#mod.request_uri, "Not found"}}
-                               | ModData#mod.data]};
-		ResponseResult ->
-		    ResponseResult
-	    end;
-	false ->
-	    {proceed, [{status, {403, ModData#mod.request_uri,
-				 ?NICE("Client not authorized to evaluate: "
-				       ++  ESIBody)}} | ModData#mod.data]}
-    end;
-generate_webpage(ModData, ESIBody, Modules, ModuleName, FunctionName,
-		 Input, ScriptElements) ->
-    try
-        Module = list_to_existing_atom(ModuleName),
-        _ = code:ensure_loaded(Module),
-        Function = list_to_existing_atom(FunctionName),
-        generate_webpage(ModData, ESIBody, Modules, Module, Function,
-                         Input, ScriptElements)
-    catch
-        _:_ ->
-            {proceed, [{status, {404, ModData#mod.request_uri, "Not found"}}
-                      | ModData#mod.data]}
+generate_webpage(ModData, ESIBody, AllowedModules0, ModuleName, FunctionName,
+		 Input, ScriptElements)
+  when is_list(ModuleName), is_list(FunctionName) ->
+    Error404 =
+        [{status, {404, ModData#mod.request_uri, "Not found"}} |
+         ModData#mod.data],
+    ConversionResult =
+        try
+            Module = list_to_existing_atom(ModuleName),
+            _ = code:ensure_loaded(Module),
+            Function = list_to_existing_atom(FunctionName),
+            {ok, Module, Function}
+        catch
+            _:_ ->
+                {proceed, Error404}
+        end,
+    case ConversionResult of
+        {ok, M, F} ->
+            AllowedModules = case AllowedModules0 of
+                                 [all] -> [M];
+                                 _ -> AllowedModules0
+                             end,
+            case lists:member(M, AllowedModules) of
+                true ->
+                    Env = httpd_script_env:create_env(esi, ModData, ScriptElements),
+                    case erl_scheme_webpage_chunk(M, F, Env, Input, ModData) of
+                        {error, erl_scheme_webpage_chunk_undefined} ->
+                            {proceed, Error404};
+                        ResponseResult ->
+                            ResponseResult
+                    end;
+                false ->
+                    Error403 =
+                        [{status,
+                          {403, ModData#mod.request_uri,
+                           ?NICE("Client not authorized to evaluate: " ++ ESIBody)}} |
+                         ModData#mod.data],
+                    {proceed, Error403}
+            end;
+        _ ->
+            ConversionResult
```

Comment:
Sorry but I do not think this makes it more readable,  just unreadable in an other way! I think we should keep two function clauses and make a new function for line 245 to 263.  As we discussed earlier it was probably written recursively after formatting to avoid coming up with a new function name.  We could call the new function do_genraret_webpage to wok around the problem.

Reply:
please check new proposal based on discussion